service-deployment/bootstrap/prometheus/prometheus-configmap.yaml.template

```diff
@@ -23,18 +23,24 @@ data:
   prometheus.yml: |-
     # Scrape config for cluster components.
     scrape_configs: 
-    - job_name: "node_exporter"
-      scrape_interval: "15s"
-      static_configs:
-      - targets: 
-        {% for host in machinelist %}
-        - {{ machinelist[host]['ip']}}:{{ clusterinfo['prometheusinfo']['node_exporter_port'] }}
-        {% endfor %}
-    - job_name: "cadvisor"
-      scrape_interval: "15s"
-      static_configs:
-      - targets: 
-        {% for host in machinelist %}
-        - {{ machinelist[host]['ip']}}:{{ clusterinfo['prometheusinfo']['cadvisor_port'] }}
-        {% endfor %}
+    - job_name: 'node_exporter'
+      scrape_interval: 30s
+      kubernetes_sd_configs:
+      - api_server: '{{ clusterinfo['webportalinfo']['k8s_api_server_uri'] }}'
+        role: node
+      relabel_configs:
+      - source_labels: [__address__]
+        regex: '(.*):(.*)'
```

Comment:
>:(.*) [](start = 20, length = 5)

You can makes this part optional such as (.*)[:\d+] so that whatever __address__ contains port or not, you just pick up the host part.

Reply:
For example, if ":" does not exists in __address__, the matching will fail thus the real ports will not be added.

---
In reply to: [178004678](https://github.com/Microsoft/pai/pull/395#discussion_r178004678) [](ancestors = 178004678)

Reply:
(.*)[:\d+] cannot match hostip case.
Assume we match 3 test case:
result1 = p.findall('127.0.0.1:1250')
result2 = p.findall('127.0.0.1:')
result3 = p.findall('127.0.0.1')

(.*)[:\d+] will cut result3's tail. 
['127.0.0.1:125']
['127.0.0.1']
['127.0.0.']

Other regex is also meet the same problem, hostip case is so flexible and difficult to combine with hostip: together to match. 

Suggest we could have another direction to think this. How about we only match the IP, this is easy to match the 3 cases.

Reply:
Solution2: use ip regex to extract the ip address and ignore other tag
p = re.compile(r'((?:(?:25[0-5]|2[0-4]\d|((1\d{2})|([1-9]?\d)))\.){3}(?:25[0-5]|2[0-4]\d|((1\d{2})|([1-9]?\d))))')

result1 = p.findall('127.0.0.1:1250')
result2 = p.findall('127.0.0.1:')
result3 = p.findall('127.0.0.1')

[('127.0.0.1', '0', '127', '0', '1', '', '1')]
[('127.0.0.1', '0', '127', '0', '1', '', '1')]
[('127.0.0.1', '0', '127', '0', '1', '', '1')]

Reply:
I found this regx works:
([^:]+)(:\d*)?

---
In reply to: [178223066](https://github.com/Microsoft/pai/pull/395#discussion_r178223066) [](ancestors = 178223066)